### ISSUE TITLE
Support arrayLength for fixed-size arrays

### ISSUE THREAD

**Issue by ben-clayton** [gpuweb/gpuweb#2074]
Following on from discussions in https://github.com/gpuweb/gpuweb/issues/1782#issuecomment-907358085 and the WebGPU matrix chat, it has been suggested by @kainino0x  that type aliases may offer a viable alternative to supporting array size as a `constexpr` (https://github.com/gpuweb/gpuweb/issues/1272).

@munrocket wrote in the WebGPU matrix chat:

```rust
type MyArray= array<vec2<f32>, 10>>;

// still need size somewhere 
var<private> MySize = 10;
```

Currently there is no way to retrieve the array size for a *fixed-size* array, so iterating over the elements of an array with a `for` loop will require the array-length to be declared at least twice (once for the `type MyArray`, once for the `for` loop upper bound).

We could change `arrayLength()` to support both fixed-size and runtime-sized arrays. If we did this, then the `MyArray` type alias can be the single source of truth for the array length.

**Comment by kvark** [comment]
See also - https://github.com/gpuweb/gpuweb/pull/1121#discussion_r497710116 from a year ago :)

**Comment by dneto0** [comment]
A year ago I was in a minimalist mood.  Now it's supported by the additional feedback.

This was briefly discussed in today's WGSL meeting. [notes](https://docs.google.com/document/d/1w8vmiIJ9QOpPT2Kf58BQ93HHGfH44TszPLA6XbrcQ50/edit). It was supported in general.  

* I argued it should support the same pointer-to-array form (but sized array).  This goes along with the restriction that general indexing into an array is only possible on a reference-to-array, so you already can form the pointer to it.  
* @kvark  noted that this functionality should be supported for arrays in uniform storage.
    * I agree, but that will need an extra carve-out in the "restrictions on functions" section. That's fine.

* @litherum also argued it should also work where the parameter is an array-value, for generality, sake.  
   * Discussion ensued as to whether the programmer really wants the array-value to be *evaluated* before invoking the builtin function.
       * I suggested no, in analogy to C's sizeof, and to avoid spurious and useless copies.  But this would require a carveout in the spec for this one builtin function.
       * @litherum suggested the argument should be evaluated as any other function call.

**Comment by jrprice** [comment]
> so iterating over the elements of an array with a for loop will require the array-length to be declared at least twice (once for the type MyArray, once for the for loop upper bound)

I'm not sure I agree with this. Now that array sizes can be module-scope constants, you could just as easily write code like this:

```
let MySize = 10;
...
{
  var myarray : array<f32, MySize>;
  for (var i = 0; i < MySize; i++) {
    // do stuff with myarray[i]
}
```

I'm not especially opposed to adding `arrayLength()` overloads for fixed-size arrays, but I'm not sure the example given in this issue is a motivating use-case.

**Comment by ben-clayton** [comment]
You're right. I missed the spec change in https://github.com/gpuweb/gpuweb/pull/2031, and only became aware of this with our discussion yesterday.
With that, declaring an array-size-from-let would be a more sensible way to have a single declaration for the size.

**Comment by munrocket** [comment]
Ben Clayton thank you for creating this issue.

@jrprice it is the first thing that I tried. The only difference that I am passed this array into function. Sure that there are different ways how to solve this.

For example arrayLength() is the only way solve it when you don’t know exact array size in *generalised* function. But module-scope constants is the thing that was in GLSL for fixed size arrays.

**Comment by ben-clayton** [comment]
> For example arrayLength() is the only way solve it when you don’t know exact array size in _generalised_ function. But module-scope constants is the thing that was in GLSL for fixed size arrays.

Please can you define "_generalised_ function" ? WGSL does not support generics for user-declared functions, and the array size is part of the array type. Given that, there currently is no way to generalize a function for different array sizes, unless you use something like a preprocessor / code-generator.

**Comment by greggman** [comment]
bump...

I'm running into this issue too. Example

```
        let color = array(
          vec4f(1, 0, 0, 1), // red
          vec4f(1, 1, 0, 1), // yellow
          vec4f(0, 1, 0, 1), // green
        );
        return color[triangleIndex % 3];  // <- have to manually specify this 3
```

wish I could do one of

```
        return color[triangleIndex % arraysize(color)]; 
```

or

```
        return color[triangleIndex % arraysize(typeof(color))]; 
```


or

```
        return color[triangleIndex % color.length]; 
```

or

```
        return color[triangleIndex % typeof(color).length]; 
```

or

```
        return color[triangleIndex % arrayLength<typeof(color)>()];
```

or anything really

**Comment by dj2** [comment]
It seems like `arrayLength(colour)` would best fit with what we have currently, then it matches the runtime `arrayLength` call. Having to call it one way for runtime and a different way for fixed size is going to be a source of confusion.

In both cases, it's returning the number of elements in the array, even if in one case it's also part of the type.

**Comment by ben-clayton** [comment]
@dj2  - see https://github.com/gpuweb/gpuweb/pull/2081#issuecomment-915052204.

**Comment by ben-clayton** [comment]
> Should we give a harder look (relook?) at `.length` and deprecate `arrayLength` for runtime-arrays? Then it would be a consistent `foo.length` call in both cases.

I would be happy with `expr.length` where `expr` could be either:
* a value-expression of a runtime or fixed-length array (e.g. `const colors = array(red, green, blue); const N = colors.length`)
* a type-expression of a runtime or fixed-length array (e.g. `alias Vertices = array<Vertex, 42>; const N = Vertices.length`)

The same type-member feature could be used to solve https://github.com/gpuweb/gpuweb/issues/3431, where we could add `f32.lowest`, `f32.highest`, etc.

**Comment by Alphapage** [comment]
Wow !!!
More than 2 years and no way to compare fixed size array length and runtime size array length...
wgsl is far from being usable for a little more advanced programming.

**Comment by Kbz-8** [comment]
up !
